README
NAME
    Harbinger::Client - Impend all the doom you could ever want â

VERSION
    version 0.001001

SYNOPSIS
     my $client = Harbinger::Client->new(
       harbinger_ip => '10.6.1.6',
       harbinger_port => 8090,
       default_args => [
         server => 'foo.lan.bar.com',
         port   => 1890,
       ],
     );

     my $doom = $client->start(
        ident => 'process-images',
     );

     for (@images) {
        ...
        $doom->bode_ill;
     }

     $client->send($doom->finish);

DESCRIPTION
    After reading The Mature Optimization Handbook
    <http://carlos.bueno.org/optimization/mature-optimization.pdf>, in a
    fever dream of hubris, I wrote "Harbinger::Client" and Harbinger::Server
    <https://github.com/frioux/Harbinger>. They have both served me
    surprisingly well with how minimal they are. The goal is to be as
    lightweight as possible such that the measuring of performance does not
    degrade performance nor impact reliability. If the client ever throws an
    exception I have failed in my goals.

    As should be clear in the "SYNOPSIS" the grim measurement that the
    "Harbinger" records is called "DOOM 💀". "DOOM 💀" currently measures a
    handful of data points, but the important bits are:

    * time

    * space

    * queries

    See more in "DOOM 💀".

   METHODS
   "new"
    Instantiate client with this method. Note example in "SYNOPSIS".

    Takes a hash of "harbinger_ip" (default of 127.0.0.1), "harbinger_port"
    (default of 8001), and "default_args" (default of "[]").

    "harbinger_ip" and "harbinger_port" are how to connect to the remote
    "HarBinger::Server".

    "default_args" get used in "start" and "instant" when forging new "DOOM
    💀".

   "start"
    The typical way to start measuring some "DOOM 💀". Note example in
    "SYNOPSIS".

    Actual implementation at "Harbinger::Client::Doom->start".

   "instant"
     $client->instant(
        ident => 'man overboard',
        count => 1,
     );

    Instead of measuring deltas as "DOOM 💀" typically does, this method is
    for measuring instantaneous events, maybe for counting or graphing them
    later. Sends the event immediately.

   "send"
     $client->send($completed_doom);

    Once "DOOM 💀" is ready to be sent to the server pass it to "send".

LIGHTHOUSE ⛯
    Beware the siren song (👄) of the Harbinger! The API is not stable yet, I
    already have major changes planned for a plugin (🔌) system. I'm not even
    going to attempt to keep things working. You've been warned (⚠).

DOOM 💀
    Measure the crushing weight, the glacial pace, the incredible demand
    which your application puts upon your database server with "DOOM™"

  DOOMFUL ATTRIBUTES ☠
   "server"
    Something unique that identifies the machine that we are measuring the
    "DOOM 💀" for. A good idea is the ip address or the hostname. If this is
    not set "DOOM 💀" will not be sent or recorded.

   "ident"
    Something unique that identifies the task that we are measuring the
    "DOOM 💀" for. For a web server, "PATH_INFO" might be a good option, or
    for some kind of message queue the task type would be a good option.

   "pid"
    The pid of the process "DOOM 💀" is being recorded for. Has a sensible
    default, you probably will never need to set it.

   "port"
    The port that the service is listening on, if applicable. Leave alone if
    unknown or not applicable.

   "count"
    The count of things being done in this unit of "DOOM 💀". If it were a
    web request that returns a list of items, this would reasonably be set
    as that number. If the operation is not related to things that are
    countable, leave alone.

   "milliseconds_elapsed"
    The total milliseconds elapsed during the unit of "DOOM 💀". If instant
    or unknown "DOOM 💀" leave empty.

   "db_query_count"
    The total queries executed during the unit of "DOOM 💀". If not
    applicable or unknown "DOOM 💀" leave empty.

   "memory_growth_in_kb"
    The total memory growth in kb during the unit of "DOOM 💀". If not
    applicable or unknown "DOOM 💀" leave empty.

   "query_logger"
    A tool to measure query count with "DBIx::Class". Please only use as
    documented, underlying implementation may change. See "QUERYLOG 📜"

  DOOMFUL METHODS 🔮
   "Harbinger::Client::Doom->start"
    Normally called via "start". Sets up some internal stuff to make
    automatic measuring of "memory_growth_in_kb" and "milliseconds_elapsed"
    work. Takes a hash and merges hash into the object via accessors.

    NOTE: to automatically measure memory growth you need either
    Win32::Process::Memory or Proc::ProcessTable installed.

   "$doom->bode_ill"
    Increment the "DOOM 💀" "count"er.

   "$doom->finish"
     $doom->finish( count => 32 );

    Finalizes "memory_growth_in_kb" and "milliseconds_elapsed". As with
    "Harbinger::Client::Doom->start" takes a hash and merges it into the
    object via accessors. Returns the object to allow chaining.

"Plack::Middleware::Harbinger"
     builder {
       enable Harbinger => {
          harbinger_ip   => '192.168.1.1',
          harbinger_port => 2250,
          default_args   => [
             server => '192.168.1.2',
             port   => 80,
          ],
       };
       $my_app
     };

    Takes the same args as "new". Adds "query_log" from "DOOM 💀" to
    "harbinger.querylog" in "psgi ENV". See "QUERYLOG 📜".

    After the query completes the "DOOM 💀" will automatically be sent.

    If "harbinger.ident" is set it will be used for the "ident", otherwise
    "PATH_INFO" will be used.

    "harbinger.server", and "harbinger.count" are passed more or less
    directly.

    "harbinger.port" will be passed if true, otherwise "SERVER_PORT" will be
    used.

"Catalyst::TraitFor::Controller::Harbinger"
    This page intentionally left blank.

QUERYLOG 📜
    You are recommended to apply the query log with
    DBIx::Class::QueryLog::Tee and DBIx::Class::QueryLog::Conditional.

    First, set up your schema package MyApp::Schema;

     use base 'DBIx::Class::Schema';
     use aliased 'DBIx::Class::QueryLog::Tee';
     use aliased 'DBIx::Class::QueryLog::Conditional';

     __PACKAGE__->load_namespaces(
        default_resultset_class => 'ResultSet',
     );

     sub connection {
        my $self = shift;

        my $ret = $self->next::method(@_);

        $ret->storage->debugobj(
           Tee->new(
              loggers => {
                 original => Conditional->new(
                    logger => $self->storage->debugobj,
                    enabled_method => sub { $ENV{DBIC_TRACE} },
                 ),
              },
           )
        );

        $ret->storage->debug(1);

        $ret
     }

     1;

    Note that the DBIx::Class::QueryLog::Tee extension allows you to add
    more Query loggers as you go, so you can even log inner loops and outer
    loops at the same time. Also note that
    DBIx::Class::QueryLog::Conditional allows you to have the "Harbinger"
    loggers always on, but the pretty DBIx::Class console logger can still
    be set via environment variable, as usual.

    Now to set the logger after whipping up some "DOOM 💀" this is all that's
    needed:

     my $doom = $client->start(
        ident => 'process-images',
     );

     $schema->storage->debugobj
       ->add_logger('process-images-harbinger', $doom->query_logger);

     $client->send($doom->finish);
     $schema->storage->debugobj
       ->remove_logger('process-images-harbinger');

    Finally, if you have some legacy code or are using the wrong ORM, you
    can still use the QueryLogger as follows:

     $dbh->{Callbacks}{ChildCallbacks}{execute} = sub {
       $doom->query_log->query_start('', []);
       $doom->query_log->query_end('', []);
       return ();
     }

    If you can pull it off, doing this dynamically with "local" is
    preferred, but that's not always possible.

AUTHOR
    Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Arthur Axel "fREW" Schmidt.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Harbinger-Client

0.001001  2014-11-20 18:02:21-06:00 America/Chicago
 - Fix a warning that did not respect the HARBINGER_WARNINGS flag
  (thanks Wes Malone!)


0.001000  2014-03-25 18:33:43-05:00 America/Chicago
 - Initial Release


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.016001;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Impend all the doom you could ever want \x{2694}",
  "AUTHOR" => "Arthur Axel \"fREW\" Schmidt <frioux+cpan\@gmail.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Harbinger-Client",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Harbinger::Client",
  "PREREQ_PM" => {
    "List::Util" => 0,
    "Module::Runtime" => 0,
    "Moo" => 0,
    "Sereal::Encoder" => 0,
    "Time::HiRes" => 0,
    "Try::Tiny" => 0,
    "namespace::clean" => 0
  },
  "VERSION" => "0.001001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "List::Util" => 0,
  "Module::Runtime" => 0,
  "Moo" => 0,
  "Sereal::Encoder" => 0,
  "Time::HiRes" => 0,
  "Try::Tiny" => 0,
  "namespace::clean" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Harbinger/Client/Doom.pm
package Harbinger::Client::Doom;
$Harbinger::Client::Doom::VERSION = '0.001001';
use List::Util 'first';
use Module::Runtime 'use_module';
use Sereal::Encoder 'encode_sereal';
use Try::Tiny;
use Time::HiRes;
use Moo;
use warnings NONFATAL => 'all';

sub _measure_memory {
   my $pid = shift;

   my $ret = try {
      if ($^O eq 'MSWin32') {
         use_module('Win32::Process::Memory')
            ->new({ pid  => $pid })
            ->get_memtotal
      } else {
         (
            first { $_->pid == $pid } @{
               use_module('Proc::ProcessTable')
               ->new
               ->table
            }
         )->rss
      }
   } catch { 0 };

   int($ret / 1024)
}

use namespace::clean;

has pid => (
   is => 'rw',
   default => sub { $$ },
);

has [qw(
   server ident
   count port milliseconds_elapsed db_query_count memory_growth_in_kb
   _start_time _start_kb query_logger
)] => ( is => 'rw' );

sub bode_ill { $_[0]->count($_[0]->count||0 + 1) }

sub start {
   my ($self, @args) = @_;

   shift->new({
      _start_time => [ Time::HiRes::gettimeofday ],
      _start_kb => _measure_memory($$),
      query_logger => use_module('DBIx::Class::QueryLog')->new,
      @args,
   })
}

sub finish {
   my ($self, %args) = @_;

   $self->milliseconds_elapsed(
      int(Time::HiRes::tv_interval($self->_start_time) * 1000)
   );
   $self->db_query_count($self->query_logger->count);
   $self->memory_growth_in_kb(_measure_memory($self->pid) - $self->_start_kb);
   $self->$_($args{$_}) for keys %args;

   return $self
}

my %mapping = (
   port                 => 'port',
   milliseconds_elapsed => 'ms',
   db_query_count       => 'qc',
   memory_growth_in_kb  => 'mg',
   count                => 'c',
);
sub _as_sereal {
   my $self = shift;

   for my $thing (qw(server ident pid)) {
      unless ($self->$thing) {
         warn "$thing can't be blank" if $ENV{HARBINGER_WARNINGS};
         return
      }
   }

   return encode_sereal({
      server => $self->server,
      ident  => $self->ident,
      pid    => $self->pid,

      map {
         my $m = $mapping{$_};
         defined $self->$_ ? ( $m => 0 + $self->$_ ) : ()
      } keys %mapping
   })
}

1;


lib/Plack/Middleware/Harbinger.pm
package Plack::Middleware::Harbinger;
$Plack::Middleware::Harbinger::VERSION = '0.001001';
use Moo;
use warnings NONFATAL => 'all';

extends 'Plack::Middleware';
use Harbinger::Client;
use namespace::clean;
use IO::Socket::INET;

has _harbinger_client => (
   is => 'ro',
   lazy => 1,
   builder => sub {
      my $self = shift;

      Harbinger::Client->new(
         harbinger_ip => $self->_harbinger_ip,
         harbinger_port => $self->_harbinger_port,
         default_args => $self->_default_args,
      )
   },
);

has _harbinger_ip => (
   is => 'ro',
   default => '127.0.0.1',
   init_arg => 'harbinger_ip',
);

has _harbinger_port => (
   is => 'ro',
   default => '8001',
   init_arg => 'harbinger_port',
);

has _default_args => (
   is => 'ro',
   default => sub { [] },
   init_arg => 'default_args',
);

has _udp_handle => (
   is => 'ro',
   builder => sub {
      IO::Socket::INET->new(
         PeerAddr => $_[0]->_harbinger_ip,
         PeerPort => $_[0]->_harbinger_port,
         Proto => 'udp'
      ) or die "couldn't connect to socket: $@" # might make this not so lethal
   },
);

sub call {
   my ($self, $env) = @_;

   my $doom = $self->_harbinger_client->start;
   # this needs to somehow pass through / wrap the other logger too
   $env->{'harbinger.querylog'} = $doom->query_logger;
   my $res = $self->app->($env);

   $self->response_cb($res, sub {
      my $ident = $env->{'harbinger.ident'} || $env->{PATH_INFO};
      $ident = "/$ident" unless $ident =~ m(^/);
      $doom->finish(
         server => $env->{'harbinger.server'},
         ident  => $ident,
         port   => $env->{'harbinger.port'} || $env->{SERVER_PORT},
         count  => $env->{'harbinger.count'},
      );

      $self->_harbinger_client->send($doom)
   })
}

1;


lib/Harbinger/Client.pm
package Harbinger::Client;
$Harbinger::Client::VERSION = '0.001001';
# ABSTRACT: Impend all the doom you could ever want ⚔

use v5.16.1;
use utf8;
use Moo;
use warnings NONFATAL => 'all';
use Try::Tiny;

use Harbinger::Client::Doom;
use IO::Socket::INET;

has _harbinger_ip => (
   is => 'ro',
   default => '127.0.0.1',
   init_arg => 'harbinger_ip',
);

has _harbinger_port => (
   is => 'ro',
   default => '8001',
   init_arg => 'harbinger_port',
);

has _udp_handle => (
   is => 'ro',
   lazy => 1,
   builder => sub {
      IO::Socket::INET->new(
         PeerAddr => $_[0]->_harbinger_ip,
         PeerPort => $_[0]->_harbinger_port,
         Proto => 'udp'
      ) or $ENV{HARBINGER_WARNINGS} && warn "couldn't connect to socket: $@"
   },
);

has _default_args => (
   is => 'ro',
   default => sub { [] },
   init_arg => 'default_args',
);

sub start {
   my $self = shift;

   Harbinger::Client::Doom->start(
      @{$self->_default_args},
      @_,
   )
}

sub instant {
   my $self = shift;

   $self->send(
      Harbinger::Client::Doom->new(
         @{$self->_default_args},
         @_,
      )
   )
}

sub send {
   my ($self, $doom) = @_;

   return unless
      my $msg = $doom->_as_sereal;

   no warnings;
   &try(sub{
      send($self->_udp_handle, $msg, 0) == length($msg)
         or $ENV{HARBINGER_WARNINGS} && warn "cannot send to: $!";
   },($ENV{HARBINGER_WARNINGS}?(catch {
      warn $_;
   }):()));
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Harbinger::Client - Impend all the doom you could ever want â

=head1 VERSION

version 0.001001

=head1 SYNOPSIS

 my $client = Harbinger::Client->new(
   harbinger_ip => '10.6.1.6',
   harbinger_port => 8090,
   default_args => [
     server => 'foo.lan.bar.com',
     port   => 1890,
   ],
 );

 my $doom = $client->start(
    ident => 'process-images',
 );

 for (@images) {
    ...
    $doom->bode_ill;
 }

 $client->send($doom->finish);

=head1 DESCRIPTION

After reading
L<The Mature Optimization Handbook|http://carlos.bueno.org/optimization/mature-optimization.pdf>,
in a fever dream of hubris, I wrote C<Harbinger::Client> and
L<Harbinger::Server|https://github.com/frioux/Harbinger>.  They have both
served me surprisingly well with how minimal they are.  The goal is to be as
lightweight as possible such that the measuring of performance does not degrade
performance nor impact reliability.  If the client B<ever> throws an exception
I have failed in my goals.

As should be clear in the L</SYNOPSIS> the grim measurement that the
C<Harbinger> records is called L</DOOM 💀>.  L</DOOM 💀> currently measures a handful
of data points, but the important bits are:

=over 2

=item * time

=item * space

=item * queries

=back

See more in L</DOOM 💀>.

=head3 METHODS

=head3 C<new>

Instantiate client with this method.  Note example in L</SYNOPSIS>.

Takes a hash of C<harbinger_ip> (default of C<127.0.0.1>), C<harbinger_port>
(default of C<8001>), and C<default_args> (default of C<[]>).

C<harbinger_ip> and C<harbinger_port> are how to connect to the remote
C<HarBinger::Server>.

C<default_args> get used in L</start> and L</instant> when forging new L</DOOM 💀>.

=head3 C<start>

The typical way to start measuring some L</DOOM 💀>.  Note example in L</SYNOPSIS>.

Actual implementation at L<< /Harbinger::Client::Doom->start >>.

=head3 C<instant>

 $client->instant(
    ident => 'man overboard',
    count => 1,
 );

Instead of measuring deltas as L</DOOM 💀> typically does, this method is for measuring
instantaneous events, maybe for counting or graphing them later.  Sends the
event immediately.

=head3 C<send>

 $client->send($completed_doom);

Once L</DOOM 💀> is ready to be sent to the server pass it to C<send>.

=head1 LIGHTHOUSE ⛯

Beware the siren song (👄) of the B<Harbinger>!  The API is not stable yet, I already
have major changes planned for a plugin (🔌) system.  I'm not even going to attempt
to keep things working.  You've been warned (⚠).

=head1 DOOM 💀

Measure the crushing weight, the glacial pace, the incredible demand which your
application puts upon your database server with C<DOOM™>

=head2 DOOMFUL ATTRIBUTES ☠

=head3 C<server>

Something unique that identifies the machine that we are measuring the L</DOOM 💀>
for.  A good idea is the ip address or the hostname.  If this is not set L</DOOM 💀>
will not be sent or recorded.

=head3 C<ident>

Something unique that identifies the task that we are measuring the L</DOOM 💀> for.
For a web server, C<PATH_INFO> might be a good option, or for some kind of
message queue the task type would be a good option.

=head3 C<pid>

The pid of the process L</DOOM 💀> is being recorded for.  Has a sensible default,
you probably will never need to set it.

=head3 C<port>

The port that the service is listening on, if applicable.  Leave alone if
unknown or not applicable.

=head3 C<count>

The count of things being done in this unit of L</DOOM 💀>.  If it were a web
request that returns a list of items, this would reasonably be set as that
number.  If the operation is not related to things that are countable, leave
alone.

=head3 C<milliseconds_elapsed>

The total milliseconds elapsed during the unit of L</DOOM 💀>.  If instant or
unknown L</DOOM 💀> leave empty.

=head3 C<db_query_count>

The total queries executed during the unit of L</DOOM 💀>.  If not applicable or
unknown L</DOOM 💀> leave empty.

=head3 C<memory_growth_in_kb>

The total memory growth in kb during the unit of L</DOOM 💀>.  If not applicable or
unknown L</DOOM 💀> leave empty.

=head3 C<query_logger>

A tool to measure query count with C<DBIx::Class>.  Please only use as
documented, underlying implementation may change.  See L</QUERYLOG 📜>

=head2 DOOMFUL METHODS 🔮

=head3 C<< Harbinger::Client::Doom->start >>

Normally called via L</start>.  Sets up some internal stuff to make automatic
measuring of L</memory_growth_in_kb> and L</milliseconds_elapsed> work.  Takes a
hash and merges hash into the object via accessors.

B<NOTE>: to automatically measure memory growth you need either
L<Win32::Process::Memory> or L<Proc::ProcessTable> installed.

=head3 C<< $doom->bode_ill >>

Increment the L</DOOM 💀> L</count>er.

=head3 C<< $doom->finish >>

 $doom->finish( count => 32 );

Finalizes L</memory_growth_in_kb> and L</milliseconds_elapsed>.  As with
L<< /Harbinger::Client::Doom->start >> takes a hash and merges it into the object
via accessors.  Returns the object to allow chaining.

=head1 C<Plack::Middleware::Harbinger>

 builder {
   enable Harbinger => {
      harbinger_ip   => '192.168.1.1',
      harbinger_port => 2250,
      default_args   => [
         server => '192.168.1.2',
         port   => 80,
      ],
   };
   $my_app
 };

Takes the same args as L</new>.  Adds C<query_log> from L</DOOM 💀> to
C<harbinger.querylog> in C<psgi ENV>.  See L</QUERYLOG 📜>.

After the query completes the L</DOOM 💀> will automatically be sent.

If C<harbinger.ident> is set it will be used for the L</ident>, otherwise
C<PATH_INFO> will be used.

C<harbinger.server>, and C<harbinger.count> are passed more or less directly.

C<harbinger.port> will be passed if true, otherwise C<SERVER_PORT> will be used.

=head1 C<Catalyst::TraitFor::Controller::Harbinger>

This page intentionally left blank.

=head1 QUERYLOG 📜

You are recommended to apply the query log with L<DBIx::Class::QueryLog::Tee>
and L<DBIx::Class::QueryLog::Conditional>.

First, set up your schema
 package MyApp::Schema;

 use base 'DBIx::Class::Schema';
 use aliased 'DBIx::Class::QueryLog::Tee';
 use aliased 'DBIx::Class::QueryLog::Conditional';

 __PACKAGE__->load_namespaces(
    default_resultset_class => 'ResultSet',
 );

 sub connection {
    my $self = shift;

    my $ret = $self->next::method(@_);

    $ret->storage->debugobj(
       Tee->new(
          loggers => {
             original => Conditional->new(
                logger => $self->storage->debugobj,
                enabled_method => sub { $ENV{DBIC_TRACE} },
             ),
          },
       )
    );

    $ret->storage->debug(1);

    $ret
 }

 1;

Note that the L<DBIx::Class::QueryLog::Tee> extension allows you to add more
Query loggers as you go, so you can even log inner loops and outer loops at the
same time.  Also note that L<DBIx::Class::QueryLog::Conditional> allows you to
have the C<Harbinger> loggers always on, but the pretty L<DBIx::Class> console
logger can still be set via environment variable, as usual.

Now to set the logger after whipping up some L</DOOM 💀> this is all that's needed:

 my $doom = $client->start(
    ident => 'process-images',
 );

 $schema->storage->debugobj
   ->add_logger('process-images-harbinger', $doom->query_logger);

 $client->send($doom->finish);
 $schema->storage->debugobj
   ->remove_logger('process-images-harbinger');

Finally, if you have some legacy code or are using the wrong ORM, you can still
use the QueryLogger as follows:

 $dbh->{Callbacks}{ChildCallbacks}{execute} = sub {
   $doom->query_log->query_start('', []);
   $doom->query_log->query_end('', []);
   return ();
 }

If you can pull it off, doing this dynamically with C<local> is preferred, but
that's not always possible.

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
